target/linux/realtek/patches-5.10/315-irqchip-irq-realtek-rtl-add-VPE-support.patch

```diff
@@ -0,0 +1,393 @@
+ {
+ 	unsigned long flags;
+ 	u32 value;
++	int cpu;
+ 
+ 	raw_spin_lock_irqsave(&irq_lock, flags);
+ 
+-	value = readl(REG(RTL_ICTL_GIMR));
+-	value |= BIT(i->hwirq);
+-	writel(value, REG(RTL_ICTL_GIMR));
++	for_each_cpu(cpu, &realtek_ictl_cpu_configurable) {
++		value = readl(REG(RTL_ICTL_GIMR, cpu));
++		value |= BIT(i->hwirq);
++		writel(value, REG(RTL_ICTL_GIMR, cpu));
++	}
+ 
+ 	raw_spin_unlock_irqrestore(&irq_lock, flags);
+ }
+@@ -44,137 +86,247 @@ static void realtek_ictl_mask_irq(struct
+ {
+ 	unsigned long flags;
+ 	u32 value;
++	int cpu;
+ 
+ 	raw_spin_lock_irqsave(&irq_lock, flags);
+ 
+-	value = readl(REG(RTL_ICTL_GIMR));
+-	value &= ~BIT(i->hwirq);
+-	writel(value, REG(RTL_ICTL_GIMR));
++	for_each_cpu(cpu, &realtek_ictl_cpu_configurable) {
++		value = readl(REG(RTL_ICTL_GIMR, cpu));
++		value &= ~BIT(i->hwirq);
++		writel(value, REG(RTL_ICTL_GIMR, cpu));
++	}
+ 
+ 	raw_spin_unlock_irqrestore(&irq_lock, flags);
+ }
+ 
++static int __maybe_unused realtek_ictl_irq_affinity(struct irq_data *i,
++	const struct cpumask *dest, bool force)
++{
++	struct realtek_ictl_output *output = i->domain->host_data;
++	cpumask_t cpu_configure;
++	cpumask_t cpu_disable;
++	cpumask_t cpu_enable;
++	unsigned long flags;
++	int cpu;
++
++	raw_spin_lock_irqsave(&irq_lock, flags);
++
++	cpumask_and(&cpu_configure, cpu_present_mask, &realtek_ictl_cpu_configurable);
++
++	cpumask_and(&cpu_enable, &cpu_configure, dest);
++	cpumask_andnot(&cpu_disable, &cpu_configure, dest);
++
++	for_each_cpu(cpu, &cpu_disable)
++		write_irr(REG(RTL_ICTL_IRR0, cpu), i->hwirq, 0);
++
++	for_each_cpu(cpu, &cpu_enable)
++		write_irr(REG(RTL_ICTL_IRR0, cpu), i->hwirq, output->output_index + 1);
++
++	irq_data_update_effective_affinity(i, &cpu_enable);
++
++	raw_spin_unlock_irqrestore(&irq_lock, flags);
++
++	return IRQ_SET_MASK_OK;
++}
++
+ static struct irq_chip realtek_ictl_irq = {
+ 	.name = "realtek-rtl-intc",
+ 	.irq_mask = realtek_ictl_mask_irq,
+ 	.irq_unmask = realtek_ictl_unmask_irq,
++#ifdef CONFIG_SMP
++	.irq_set_affinity = realtek_ictl_irq_affinity,
++#endif
+ };
+ 
+ static int intc_map(struct irq_domain *d, unsigned int irq, irq_hw_number_t hw)
+ {
+-	irq_set_chip_and_handler(hw, &realtek_ictl_irq, handle_level_irq);
++	struct realtek_ictl_output *output = d->host_data;
++	unsigned long flags;
++
++	irq_set_chip_and_handler(irq, &realtek_ictl_irq, handle_level_irq);
++
++	raw_spin_lock_irqsave(&irq_lock, flags);
++
++	output->child_mask |= BIT(hw);
++	write_irr(REG(RTL_ICTL_IRR0, 0), hw, output->output_index + 1);
++
++	raw_spin_unlock_irqrestore(&irq_lock, flags);
+ 
+ 	return 0;
+ }
+ 
++static int intc_select(struct irq_domain *d, struct irq_fwspec *fwspec,
++	enum irq_domain_bus_token bus_token)
++{
++	struct realtek_ictl_output *output = d->host_data;
++	bool routed_elsewhere;
++	unsigned long flags;
++	u32 routing_old;
++	int cpu;
++
++	if (fwspec->fwnode != output->fwnode)
++		return false;
++
++	/* Original specifiers had only one parameter */
```

Comment:
FWIW, the new (two-cell) binding got a positive review upstream, but the driver changes are still under review. The CPU affinity support hasn't been submitted upstream yet, since this depends on the currently pending patches.
As far as I'm concerned, this patch can be merged as-is, but I'll replace it by whatever gets accepted upstream.
Already accepted changes are:
- https://lore.kernel.org/all/4b4936606480265db47df152f00bc2ed46340599.1641739718.git.sander@svanheule.net/
- https://lore.kernel.org/all/177b920aa8d8610615692d0e657e509f363c85ca.1641739718.git.sander@svanheule.net/
- https://lore.kernel.org/all/5082ad3cb8b4eedf55075561b93eff6570299fe1.1641739718.git.sander@svanheule.net/
Still under review:
https://lore.kernel.org/all/cover.1644165421.git.sander@svanheule.net/

Reply:
Yes, I plan to just replace this patch by all the changes accepted upstream. Otherwise this need to be split up here first to be able to add the CPU affinity changes, and it seems like people would rather see this move forward.